File: 1047.remove-all-adjacent-duplicates-in-string.cpp
/*
 * @lc app=leetcode id=1047 lang=cpp
 *
 * [1047] Remove All Adjacent Duplicates In String
 */
#include <string>

using namespace std;

// @lc code=start
class Solution {
public:
    // In C++, pop operation can be done on std::string. No need to use 
    // an auxillary std::stack.
    string removeDuplicates(string s) {
        string res = "";
        for (char& c : s) {
            if (res.size() && c == res.back())  res.pop_back();
            else  res.push_back(c);
        }
        return res;
    }
};
// @lc code=end

File: 1047.remove-all-adjacent-duplicates-in-string.py
#
# @lc app=leetcode id=1047 lang=python3
#
# [1047] Remove All Adjacent Duplicates In String
#

# @lc code=start
class Solution:
    def removeDuplicates(self, s: str) -> str:
        stack = []

        for c in s:
            if len(stack) and stack[-1] == c:
                stack.pop()
            else:
                stack.append(c)

        return "".join(stack)

# @lc code=end

